src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -407,23 +407,63 @@ private static object BindToCollection(Type type, IConfiguration config, BinderO
             }
         }
 
-        [RequiresUnreferencedCode("In case type is a Nullable<T>, cannot statically analyze what the underlying type is so its members may be trimmed.")]
-        private static object CreateInstance([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors)] Type type)
+        [RequiresUnreferencedCode(
+            "In case type is a Nullable<T>, cannot statically analyze what the underlying type is so its members may be trimmed.")]
+        private static object CreateInstance(
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors |
+                                        DynamicallyAccessedMemberTypes.NonPublicConstructors)]
+            Type type,
+            IConfiguration config,
+            BinderOptions options)
         {
             Debug.Assert(!type.IsArray);
 
-            if (type.IsAbstract)
+            if (type.IsInterface || type.IsAbstract)
             {
                 throw new InvalidOperationException(SR.Format(SR.Error_CannotActivateAbstractOrInterface, type));
             }
 
-            if (!type.IsValueType)
+            ConstructorInfo[] constructors = type.GetConstructors(BindingFlags.Public | BindingFlags.Instance);
+
+            bool hasParameterlessConstructor =
+                type.IsValueType || constructors.Any(ctor => ctor.GetParameters().Length == 0);
+
+            if (!type.IsValueType && constructors.Length == 0)
             {
-                bool hasDefaultConstructor = type.GetConstructors(DeclaredOnlyLookup).Any(ctor => ctor.IsPublic && ctor.GetParameters().Length == 0);
-                if (!hasDefaultConstructor)
+                throw new InvalidOperationException(SR.Format(SR.Error_MissingPublicInstanceConstructor, type));
+            }
+
+            if (constructors.Length > 1 && !hasParameterlessConstructor)
+            {
+                throw new InvalidOperationException(SR.Format(SR.Error_MultipleParameterizedConstructors, type));
+            }
+
+            if (constructors.Length == 1 && !hasParameterlessConstructor)
+            {
+                ConstructorInfo constructor = constructors[0];
+                ParameterInfo[] parameters = constructor.GetParameters();
+
+                if (!CanBindToTheseConstructorParameters(parameters, out string nameOfInvalidParameter))
                 {
-                    throw new InvalidOperationException(SR.Format(SR.Error_MissingParameterlessConstructor, type));
+                    throw new InvalidOperationException(SR.Format(SR.Error_CannotBindToConstructorParameter, type, nameOfInvalidParameter));
                 }
+
+
+                var properties = GetAllProperties(type);
+
+                if (!DoAllParametersHaveEquivalentProperties(parameters, properties, out string nameOfInvalidParameters))
```

Comment:
I don't think this rule is valid. The class could have a field. Or it could expose methods to get the values. Let's just bind to the ctor parameters without this check.

Reply:
I added this in response to @halter73 's [comment](https://github.com/dotnet/runtime/pull/67258#discussion_r880943936).

Revisiting that comment, and running tests against the desired implementation that we want from S.T.J, I think I've found a bug (or at least a discrepancy in the text of the exception thrown).

Reply:
Addressing your point re. **fields**, I've updated the check in the most recent commit.  I'm not proposing that this is correct or required behaviour, but it does match what S.T.J does (or at least implies that it does, based on the bug discovered that is described in the comment link above).

Reply:
For anyone reading along: `S.T.J.` is `System.Text.Json` 🙂

Reply:
T.Y.V.M.! ;)